dask/array/core.py

```diff
@@ -2981,12 +2981,28 @@ def concatenate(seq, axis=0, allow_unknown_chunksizes=False):
     """
     seq = [asarray(a) for a in seq]
 
-    n = len(seq)
-
-    if n == 0:
+    if not seq:
         raise ValueError("Need array(s) to concatenate")
 
+    # Coerce all arrays to the same type
+    # Empty arrays can impact the output dtype
+    from .utils import meta_from_array
+    metas = [getattr(s, '_meta', s) for s in seq]
```

Comment:
Same question about assuming each item in `seq` already having a `_meta` attribute

Reply:
Same answer as above. 🙂

Reply:
Even though the documentation claims `seq` should be a list of `dask.array`s, it's permissive enough to let it be a list of NumPy-like arrays, which don't have `_meta`, thus we default to `s` and ensure we compute `meta_from_array` on the following line.

Reply:
But each item in `seq` is converted to a `dask.array` when `seq = [asarray(a) for a in seq]` is executed at the top of `concatenate`, so IIUC each item in `seq` should have a `_meta` attribute

Reply:
Yeah that true. @pentschev, the behavior in this function changed with PR ( https://github.com/dask/dask/pull/4925 ). We now are forcing them all to be Dask Arrays. Does that mean they all have `_meta` or are there instances where that would still not be the case?

Reply:
Right, I forgot that it was recently changed. All `dask.array`s should have `_meta`, so we can probably drop that condition. But you can get this PR integrated and I'll open a new one to handle this case, maybe there's something else I'm not thinking of right now, so I can fix it if that's the case.

Reply:
I've gone ahead and pushed some simplifications to the `_meta` handling code here. Hopefully that addresses this concern. Let's follow-up on the new diff if not.

Reply:
Thanks for the clarification @pentschev and modifications @jakirkham. Didn't mean to nitpick things so much : )

Reply:
Not at all. The questions were reasonable and was curious myself. Things appear to be in a better state as a result.